src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -13,472 +13,528 @@ namespace nbl::asset
+        .offset = 0,
+        .size = 0
+    };
 
-        bool anyStageNotFinished = false;
-        for (auto i=0u; i<infos.size(); i++)
-        {
-            auto& introBindings = introspections[i]->descriptorSetBindings[_set];
-            if (checkedDescsCnt[i] == introBindings.size())
-                continue;
-            anyStageNotFinished = true;
-
-            const auto& introBinding = introBindings[checkedDescsCnt[i]].binding;
-            if (introBinding < binding.binding)
-                binding.binding = introBinding;
-        }
-        if (!anyStageNotFinished) //all shader stages finished
-            break;
+    SPushConstantRange current = {
+        .offset = 0,
+        .size = 0
+    };
+
+    // TODO: test
+    // run-length encode m_pushConstantBytes
+    for (uint32_t currentByteOffset = 1u; currentByteOffset < MaxPushConstantsSize; ++currentByteOffset)
+    {
+        current.stageFlags = m_pushConstantBytes[currentByteOffset].value;
+        current.offset++;
+        current.size++;
 
-        const ICPUSpecializedShader* refShader = nullptr;
-        uint32_t refIndex = ~0u;
-        core::smart_refctd_ptr<const CIntrospectionData> refIntro = nullptr;
+        if (current.stageFlags != prev.stageFlags)
         {
-            auto checkedDescsCntIt = checkedDescsCnt;
-            auto introspectionIt = introspections;
-            for (auto shader : shaders)
+            if (prev.stageFlags)
             {
-                auto& introBindings = (*introspectionIt)->descriptorSetBindings[_set];
-                if (*checkedDescsCntIt!=introBindings.size())
-                {
-                    const auto& introBinding = introBindings[*checkedDescsCntIt].binding;
-                    if (introBinding==binding.binding)
-                    {
-                        stageFlags |= shader->getStage();
-                        refIndex = (*checkedDescsCntIt)++;
-                        refIntro = *introspectionIt;
-                        refShader = shader;
-                    }
-                }
-                introspectionIt++;
-                checkedDescsCntIt++;
+                prev.size = current.offset - prev.offset;
+                tmp.push_back(prev);
             }
+            prev = current;
         }
+    }
 
-        auto& introBnd = refIntro->descriptorSetBindings[_set][refIndex];
-        binding.type = resType2descType(introBnd.type);
-        binding.stageFlags = static_cast<ICPUShader::E_SHADER_STAGE>(stageFlags);
-        binding.count = introBnd.descriptorCount;
-        if (introBnd.descCountIsSpecConstant)
-        {
-            auto& specInfo = refShader->getSpecializationInfo();
-            auto val = specInfo.getSpecializationByteValue(binding.count);
-            assert(val.second == 4ull);
-            memcpy(&binding.count, val.first, 4ull);
-        }
+    if (prev.stageFlags)
+    {
+        prev.size = current.offset - prev.offset;
```

Comment:
after reaching the end of the loop `currentByteOffset==MaxPushConstantsSize` which means taht `current.offset` will be `MaxPushCosntantsSize-1` as well

This is why you need to set the size to `MaxPushConstantsSize-prev.offset`